File: Software/apiexamples/lightpack.py
import socket, time, imaplib, re, sys

class lightpack:

#	host = '127.0.0.1'    # The remote host
#	port = 3636              # The same port as used by the server
#	ledMap = [1,2,3,4,5,6,7,8,9,10] 	#mapped LEDs
	
	def __init__(self, _host, _port, _ledMap):
		self.host = _host
		self.port = _port
		self.ledMap = _ledMap
	
	def __readResult(self):	# Return last-command API answer  (call in every local method)
		total_data=[]
		data = self.connection.recv(8192)
		total_data.append(data)
		return ''.join(total_data)
		
	def getProfiles(self):
		cmd = 'getprofiles\n'
		self.connection.send(cmd)
		profiles = self.__readResult()
		return profiles.split(':')[1].rstrip(';\n').split(';')		
		
	def getProfile(self):
		cmd = 'getprofile\n'
		self.connection.send(cmd)
		profile = self.__readResult()
		profile = profile.split(':')[1]
		return profile
		
	def getStatus(self):
		cmd = 'getstatus\n'
		self.connection.send(cmd)
		status = self.__readResult()
		status = status.split(':')[1]
		return status
		
	def getAPIStatus(self):
		cmd = 'getstatusapi\n'
		self.connection.send(cmd)
		status = self.__readResult()
		status = status.split(':')[1]
		return status
		
	def connect (self):
		try: 	#Try to connect to the server API
			self.connection = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
			self.connection.connect((self.host, self.port))			
			self.__readResult()
		except:
			print 'Lightpack API server is missing'
			sys.exit(0)
		
	def setColor(self, n, r, g, b): 	# Set color to the define LED		
		cmd = 'setcolor:{0}-{1},{2},{3}\n'.format(self.ledMap[n-1], r, g, b)
		self.connection.send(cmd)
		self.__readResult()
		
	def setColorToAll(self, r, g, b): 	# Set one color to all LEDs
		cmdstr = ''
		for i in self.ledMap:
			cmdstr = str(cmdstr) + str(i) + '-{0},{1},{2};'.format(r,g,b)		
		cmd = 'setcolor:' + cmdstr + '\n'	
		self.__readResult()
		self.connection.send(cmd)

	def setGamma(self, g):
		cmd = 'setgamma:{0}\n'.format(g)
		self.connection.send(cmd)
		self.__readResult()		
		
	def setSmooth(self, s):
		cmd = 'setsmooth:{0}\n'.format(s)
		self.connection.send(cmd)
		self.__readResult()

	def setProfile(self, p):
		cmd = 'setprofile:{0}\n'.format(p)
		self.connection.send(cmd)		
		self.__readResult()
		
	def lock(self):
		cmd = 'lock\n'
		self.connection.send(cmd)
		self.__readResult()
		
	def unlock(self):
		cmd = 'unlock\n'
		self.connection.send(cmd)	
		self.__readResult()
	
	def turnOn(self):
		cmd = 'setstatus:on\n'		
		self.connection.send(cmd)
		self.__readResult()
	
	def turnOff(self):
		cmd = 'setstatus:off\n'
		self.connection.send(cmd)
		self.__readResult()		

	def disconnect(self):
		self.unlock()
		self.connection.close()
		

File: Software/apiexamples/Ruby/lightpack.rb
# coding: utf-8

# Title:       LightPack API wrapper for Ruby
# Author:      Mikhail S. Pobolovets
# License:     GPL v3
# Date:        2011/08/15
# API Version: 5.5

require 'socket'

class LightPack
  include Socket::Constants

  #  host = '127.0.0.1'                # The remote host
  #  port = 3636                       # The same port as used by the server
  #  led_map = [1,2,3,4,5,6,7,8,9,10]  # mapped LEDs

  def initialize(host, port, led_map)
    @host = host
    @port = port
    @led_map = led_map
  end

  def connect
    # Try to connect to the API server
    @connection = Socket.new(AF_INET, SOCK_STREAM, 0)
    sockaddr = Socket.pack_sockaddr_in(@port, @host)
    @connection.connect(sockaddr)
    read_result
  rescue
    puts "Lightpack API server is missing"
    exit 0
  end

  # Return last-command API answer
  def read_result
    @connection.recv(8192).to_s
  end

  # Execute command and recieve an answer
  def exec_cmd(cmd)
    @connection.send "#{cmd}\n"
    read_result
  end

  def get_profiles
    exec_cmd('getprofiles').split(':')[1].rstrip(';\n').split(';')
  end

  def get_profile
    exec_cmd('getprofile').split(':')[1]
  end

  def get_status
    exec_cmd('getstatus').split(':')[1]
  end

  def get_api_status
    exec_cmd('getstatusapi').split(':')[1]
  end

  # Set color to the define LED
  def set_color(n, r, g, b)
    exec_cmd "setcolor:#{@led_map[n-1]}-#{r},#{g},#{b}"
  end

  # Set one color to all LEDs
  def set_color_to_all(r, g, b)
    cmd = "setcolor:"
    @led_map.each {|i| cmd += "#{i}-#{r},#{g},#{b};" }
    exec_cmd cmd
  end

  def set_gamma(g)
    exec_cmd "setgamma:#{g}"
  end

  def set_smooth(s)
    exec_cmd "setsmooth:#{s}"
  end

  def set_profile(p)
    exec_cmd "setprofile:#{p}"
  end

  def lock
    exec_cmd 'lock'
  end

  def unlock
    exec_cmd 'unlock'
  end

  def turn_on
    exec_cmd 'setstatus:on'
  end

  def turn_off
    exec_cmd 'setstatus:off'
  end

  def disconnect
    unlock
    @connection.close()
  end

end

